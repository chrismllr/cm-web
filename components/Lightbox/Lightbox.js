import React from 'react'
import MaterialIcon from '../MaterialIcon'

function commaSeparatedString (arr) {
  return arr.reduce((acc, str) => acc.length ? `${acc}, ${str}` : str, '')
}

const Lightbox = (props) => (
  <div className="Lightbox">
    <a onClick={props.close} className="Back">
      <span>&larr;</span>
    </a>

    <img className="Image" src={props.lightboxProject.img} alt={props.lightboxProject.name} />

    <div className="Lb-content">
      <div className="Col1">
        <h2>
          {props.lightboxProject.name}
          {props.lightboxProject.projectHref &&
          <a className="Project-link" href={props.lightboxProject.projectHref} target="_blank">
            <MaterialIcon icon="link" />
          </a>
        }
        </h2>
        <span className="Sub">
          {commaSeparatedString(props.lightboxProject.technologies)}
        </span>
      </div>

      <div className="Col2">
        <span className="Learnings">Learnings:</span>
        <ul>
          {props.lightboxProject.learnings.map((l, i) => <li key={i}>{l}</li>)}
        </ul>
      </div>
    </div>

    <style jsx>{`
      @keyframes fadeIn {
        0% { opacity: 0; }
        100% { opacity: 1; }
      }

      @keyframes fadeInDown {
        0% { opacity: 0; transform: translate3d(0, -50%, 0); }
        100% { opacity: 1; transform: translate3d(0, 0, 0); }
      }

      .Lightbox {
        position: fixed;
        top: 0;
        bottom: 0;
        left: 0;
        right: 0;
        background-color: rgba(255,255,107, .98);
        z-index: 100;
        padding: 2rem;

        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;

        animation: fadeIn;
        animation-duration: .2s;
        animation-fill-mode: both;
      }

      .Back {
        font-size: 2rem;
        color: #222;
        text-decoration: none;
        opacity: .8;
        cursor: pointer;
        margin-bottom: 1rem;
      }

      .Image {
        max-height: 65vh;
        max-width: 90vw;
        box-shadow: rgba(0,0,0,.1) 0 0 50px 15px;

        animation: fadeInDown;
        animation-fill-mode: both;
        animation-duration: .5s;
        animation-timing-function: cubic-bezier(0,1,.51,1);
        animation-delay: .2s;
      }

      .Lb-content {
        color: rgba(34,34,34, .8);
        letter-spacing: .03rem;
        font-size: .875rem;
        display: flex;
        align-items: flex-start;
        margin-top: 2rem;
      }

      .Lb-content,
      .Back {
        animation: fadeIn;
        animation-fill-mode: both;
        animation-duration: .5s;
        animation-delay: .4s;
      }

      .Learnings {
        font-family: "Arvo", serif;
        display: inline-block;
        margin-bottom: 1.33rem;
      }

      .Project-link {
        text-decoration: none;
        font-size: 1.2rem;
        margin-left: .5rem;
        color: #222;
        opacity: .5;
        transition: .3s ease;
      }

      .Project-link:hover {
        opacity: 1;
      }

      .Col1 {
        border-right: 1px solid rgba(34,34,34, .1);
        padding: 0 3rem 1rem 1rem;
      }

      .Col2 {
        padding: .66rem 1rem 1rem 3rem;
      }

      .Col1 h2 {
        font-size: 2.5rem;
        margin-bottom: .5rem;
      }

      .Col2 ul {
        padding: 0;
        margin: 0;
        list-style: none;
      }

      .Col2 li:not(:last-child) {
        margin-bottom: .5rem;
      }

      .Sub {
        opacity: .7;
        font-style: italic;
      }

      @media (max-width: 630px) {
        .Lb-content {
          flex-direction: column;
        }

        .Col1, .Col2 {
          padding: 1rem 0;
          border: none;
        }

        .Col1 h2 {
          font-size: 2rem;
          margin-bottom: 0;
        }

        .Learnings {
          margin-bottom: 1rem;
        }
      }
    `}</style>
  </div>
)

Lightbox.propTypes = {
  close: React.PropTypes.func,
  lightboxProject: React.PropTypes.shape({
    img: React.PropTypes.string,
    name: React.PropTypes.string,
    projectHref: React.PropTypes.string,
    technologies: React.PropTypes.arrayOf(React.PropTypes.string),
    learnings: React.PropTypes.arrayOf(React.PropTypes.string)
  })
}

export default Lightbox
