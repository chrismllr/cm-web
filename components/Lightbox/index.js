import React from 'react'
import PropTypes from 'prop-types'
import MaterialIcon from '../MaterialIcon'
import { fonts } from '../../styles/common'

function commaSeparatedString(arr) {
  return arr.reduce((acc, str) => {
    return acc.length ? `${acc}, ${str}` : str
  }, '')
}

// TODO: Styled components
function Lightbox(props) {
  return (
    <div className="Lightbox">
      <a data-hook="close" onClick={props.close} className="Back">
        <span>&larr;</span>
      </a>

      <img
        className="Image"
        src={props.lightboxProject.img}
        alt={props.lightboxProject.name}
      />

      <div className="Lb-content">
        <div className="Col1">
          <h2>
            {props.lightboxProject.name}
            {props.lightboxProject.projectHref && (
              <a
                className="Project-link"
                href={props.lightboxProject.projectHref}
                target="_blank"
                rel="noopener"
                data-hook="project-link"
              >
                <MaterialIcon icon="link" />
              </a>
            )}
          </h2>
          <p className="Company">{props.lightboxProject.company}</p>
          <p className="Sub">
            {commaSeparatedString(props.lightboxProject.technologies)}
          </p>
        </div>

        <div className="Col2">
          <span className="Learnings">Learnings:</span>
          <ul>
            {props.lightboxProject.learnings.map((l, i) => (
              <li key={i}>{l}</li>
            ))}
          </ul>
        </div>
      </div>

      <style jsx>{`
        @keyframes fadeIn {
          0% {
            opacity: 0;
          }
          100% {
            opacity: 1;
          }
        }

        @keyframes fadeInDown {
          0% {
            opacity: 0;
            transform: translate3d(0, -50%, 0);
          }
          100% {
            opacity: 1;
            transform: translate3d(0, 0, 0);
          }
        }

        .Lightbox {
          position: fixed;
          top: 0;
          bottom: 0;
          left: 0;
          right: 0;
          overflow-y: auto;
          background-color: rgba(255, 255, 255, 0.98);
          z-index: 100;
          padding: 2rem;

          animation: fadeIn;
          animation-duration: 0.2s;
          animation-fill-mode: both;

          padding-left: calc(2rem + env(safe-area-inset-left));
          padding-right: calc(2rem + env(safe-area-inset-right));
          padding-top: calc(2rem + env(safe-area-inset-top));
          padding-bottom: calc(2rem + env(safe-area-inset-bottom));
        }

        .Back {
          font-size: 1.5rem;
          color: #222;
          display: block;
          text-decoration: none;
          opacity: 0.8;
          cursor: pointer;
          margin-bottom: 1rem;
        }

        .Image {
          max-height: 55vh;
          max-width: calc(100vw - 4rem);
          box-shadow: rgba(0, 0, 0, 0.1) 0 0 50px 15px;

          animation: fadeInDown;
          animation-fill-mode: both;
          animation-duration: 0.5s;
          animation-timing-function: cubic-bezier(0, 1, 0.51, 1);
          animation-delay: 0.2s;
        }

        .Lb-content {
          color: rgba(34, 34, 34, 0.8);
          letter-spacing: 0.03rem;
          font-size: 0.875rem;
          margin: 1rem 0;
        }

        .Lb-content,
        .Back {
          animation: fadeIn;
          animation-fill-mode: both;
          animation-duration: 0.5s;
          animation-delay: 0.4s;
        }

        .Learnings {
          font-family: ${fonts.PRIMARY};
          display: inline-block;
          margin-bottom: 1rem;
        }

        .Project-link {
          text-decoration: none;
          font-size: 1.2rem;
          margin-left: 0.5rem;
          color: #222;
          opacity: 0.5;
          transition: 0.3s ease;
        }

        .Project-link:hover {
          opacity: 1;
        }

        .Col1,
        .Col2 {
          padding: 1rem 0;
          border: none;
        }

        .Col1 h2 {
          font-size: 2rem;
          margin-bottom: 0;
        }

        .Col2 ul {
          padding: 0;
          margin: 0;
          list-style: none;
        }

        .Col2 li:not(:last-child) {
          margin-bottom: 0.5rem;
        }

        .Sub {
          opacity: 0.7;
          font-style: italic;
        }

        @media screen and (min-width: 630px) {
          .Lightbox {
            display: flex;
            align-items: center;
            justify-content: center;
            flex-direction: column;
          }

          .Back {
            font-size: 2rem;
          }

          .Lb-content {
            display: flex;
            align-items: flex-start;
            margin: 2rem 0;
          }

          .Learnings {
            margin-bottom: 1.33rem;
          }

          .Col1 {
            padding: 0 3rem 1rem 1rem;
            border-right: 1px solid rgba(34, 34, 34, 0.1);
          }

          .Col2 {
            padding: 0.66rem 1rem 1rem 3rem;
          }

          .Col1 h2 {
            line-height: 1.1;
            margin-bottom: 1rem;
            font-size: 2.5rem;
        }
      `}</style>
    </div>
  )
}

Lightbox.propTypes = {
  close: PropTypes.func,
  lightboxProject: PropTypes.shape({
    img: PropTypes.string,
    name: PropTypes.string,
    company: PropTypes.string,
    projectHref: PropTypes.string,
    technologies: PropTypes.arrayOf(PropTypes.string),
    learnings: PropTypes.arrayOf(PropTypes.string)
  })
}

export default Lightbox
