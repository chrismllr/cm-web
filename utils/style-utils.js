const spacingFactor = 8
const spaceVars = {
  space0: `${spacingFactor / 2}px`,  // 4
  space1: `${spacingFactor}px`,      // 8
  space2: `${spacingFactor * 2}px`,  // 16
  space3: `${spacingFactor * 3}px`,  // 24
  space4: `${spacingFactor * 4}px`,  // 32
  space5: `${spacingFactor * 5}px`,  // 40
  space6: `${spacingFactor * 6}px`,  // 48

  space8: `${spacingFactor * 8}px`,  // 64
  space9: `${spacingFactor * 9}px`,  // 72
  space13: `${spacingFactor * 13}px` // 104
}

export function spacing (sp) {
  return spaceVars[`space${sp}`]
}
