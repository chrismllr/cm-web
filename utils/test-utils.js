export function hook (identifier) {
  return '[data-hook="' + identifier + '"]'
}
