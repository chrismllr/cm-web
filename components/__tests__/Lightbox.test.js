import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Lightbox from '../Lightbox'
import { hook } from '../../utils/test-utils'

function setup (args = {}) {
  const props = {
    close: jest.fn(),
    lightboxProject: {
      img: 'img.jpg',
      name: 'JLeveritt',
      company: 'Personal',
      projectHref: 'www.project.com',
      technologies: ['React'],
      learnings: ['Did cool stuff']
    },
    ...args
  }
  const wrapper = shallow(<Lightbox {...props} />)
  const component = renderer.create(<Lightbox {...props} />)
  const json = component.toJSON()

  return { props, wrapper, component, json }
}

describe('(Component) Lightbox', () => {
  it('Renders to snapshot', () => {
    const { json } = setup()
    expect(json).toMatchSnapshot()
  })

  it('When back arrow is clicked lightbox is closed', () => {
    const { wrapper, props } = setup()
    wrapper.find(hook('close')).simulate('click')

    expect(props.close).toHaveBeenCalled()
  })

  it('If project link exists, renders link', () => {
    const { wrapper } = setup()
    expect(wrapper.find(hook('project-link')).exists()).toBeTruthy()

    const { wrapper: w2 } = setup({
      lightboxProject: {
        img: 'img.jpg',
        name: 'JLeveritt',
        company: 'Personal',
        projectHref: undefined,
        technologies: ['React'],
        learnings: ['Did cool stuff']
      }
    })
    expect(w2.find(hook('project-link')).exists()).not.toBeTruthy()
  })
})
