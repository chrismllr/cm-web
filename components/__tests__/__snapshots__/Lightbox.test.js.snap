// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Lightbox Renders to snapshot 1`] = `
<div
  className="Lightbox"
  data-jsx={4265853809}
>
  <a
    className="Back"
    data-hook="close"
    data-jsx={4265853809}
    onClick={[Function]}
  >
    <span
      data-jsx={4265853809}
    >
      ‚Üê
    </span>
  </a>
  <img
    alt="JLeveritt"
    className="Image"
    data-jsx={4265853809}
    src="img.jpg"
  />
  <div
    className="Lb-content"
    data-jsx={4265853809}
  >
    <div
      className="Col1"
      data-jsx={4265853809}
    >
      <h2
        data-jsx={4265853809}
      >
        JLeveritt
        <a
          className="Project-link"
          data-hook="project-link"
          data-jsx={4265853809}
          href="www.project.com"
          rel="noopener"
          target="_blank"
        >
          <i
            className="icon icon-link"
            data-jsx={1566661193}
            style={undefined}
          />
        </a>
      </h2>
      <p
        className="Company"
        data-jsx={4265853809}
      >
        Personal
      </p>
      <p
        className="Sub"
        data-jsx={4265853809}
      >
        React
      </p>
    </div>
    <div
      className="Col2"
      data-jsx={4265853809}
    >
      <span
        className="Learnings"
        data-jsx={4265853809}
      >
        Learnings:
      </span>
      <ul
        data-jsx={4265853809}
      >
        <li
          data-jsx={4265853809}
        >
          Did cool stuff
        </li>
      </ul>
    </div>
  </div>
</div>
`;
